### PR TITLE
refactor: Replace `url.parse` with `new URL()`

### PR DESCRIPTION
**Pull request by xxczaki**
I think this is the last thing we need to do in order to release the 3.x alpha release.

Replacing the `url.parse` API here doesn't break our tests:
https://github.com/bitinn/node-fetch/blob/3b452b8e1f280142d78d5a33abf945d77c25ea59/src/request.js#L51

However, trying to use the new WHATWG API here:
https://github.com/bitinn/node-fetch/blob/3b452b8e1f280142d78d5a33abf945d77c25ea59/src/request.js#L54

throws some errors:

```
> node-fetch@2.6.0 test /home/akepinski/dev/node-fetch
> cross-env BABEL_ENV=test mocha --require @babel/register --throw-deprecation test/test.js



  node-fetch
    ✓ should return a promise
    1) should return a promise
(node:11807) UnhandledPromiseRejectionWarning: FetchError: request to http://localhost:30001/hello failed, reason: connect ECONNREFUSED 127.0.0.1:80
    at ClientRequest.<anonymous> (/home/akepinski/dev/node-fetch/src/index.js:113:11)
    at ClientRequest.emit (events.js:210:5)
    at Socket.socketErrorListener (_http_client.js:406:9)
    at Socket.emit (events.js:210:5)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
(node:11807) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)


  1 passing (35ms)
  1 failing

  1) node-fetch
       should return a promise:
     Uncaught DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
      at emitDeprecationWarning (internal/process/promises.js:153:11)
      at processPromiseRejections (internal/process/promises.js:205:13)
      at processTicksAndRejections (internal/process/task_queues.js:94:32)



npm ERR! Test failed.  See above for more details.
```

Trying to use the new API [here](https://github.com/bitinn/node-fetch/blob/3b452b8e1f280142d78d5a33abf945d77c25ea59/src/request.js#L59) also breaks our tests, but with more precise errors:

```
175 passing (2s)
  29 failing

  1) node-fetch
       should follow redirect code 301:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  2) node-fetch
       should follow redirect code 302:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  3) node-fetch
       should follow redirect code 303:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  4) node-fetch
       should follow redirect code 307:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  5) node-fetch
       should follow redirect code 308:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  6) node-fetch
       should follow redirect chain:
     FetchError: request to http://localhost:30001/redirect/301 failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  7) node-fetch
       should follow POST request redirect code 301 with GET:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  8) node-fetch
       should follow PATCH request redirect code 301 with PATCH:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  9) node-fetch
       should follow POST request redirect code 302 with GET:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  10) node-fetch
       should follow PATCH request redirect code 302 with PATCH:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  11) node-fetch
       should follow redirect code 303 with GET:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  12) node-fetch
       should follow PATCH request redirect code 307 with PATCH:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  13) node-fetch
       should obey maximum redirect, reject case:

      AssertionError: expected { type: 'system',
  name: 'FetchError',
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED',
  erroredSysCall: [Error: connect ECONNREFUSED 127.0.0.1:80] } to have property 'type' of 'max-redirect', but got 'system'
      + expected - actual

      -system
      +max-redirect
      
      at /home/akepinski/dev/node-fetch/node_modules/chai-as-promised/lib/chai-as-promised.js:302:22
      at processTicksAndRejections (internal/process/task_queues.js:93:5)

  14) node-fetch
       should obey redirect chain, resolve case:
     FetchError: request to http://localhost:30001/redirect/301 failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  15) node-fetch
       should follow redirect code 301 and keep existing headers:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  16) node-fetch
       should set redirected property on response when redirect:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  17) node-fetch
       should allow custom timeout on redirected requests:

      AssertionError: expected { type: 'system',
  name: 'FetchError',
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED',
  erroredSysCall: [Error: connect ECONNREFUSED 127.0.0.1:80] } to have property 'type' of 'request-timeout', but got 'system'
      + expected - actual

      -system
      +request-timeout
      
      at /home/akepinski/dev/node-fetch/node_modules/chai-as-promised/lib/chai-as-promised.js:302:22
      at processTicksAndRejections (internal/process/task_queues.js:93:5)

  18) node-fetch
       should allow redirects to be aborted:

      AssertionError: expected { type: 'system',
  name: 'FetchError',
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED',
  erroredSysCall: [Error: connect ECONNREFUSED 127.0.0.1:80] } to have property 'name' of 'AbortError', but got 'FetchError'
      + expected - actual

      -FetchError
      +AbortError
      
      at /home/akepinski/dev/node-fetch/node_modules/chai-as-promised/lib/chai-as-promised.js:302:22
      at processTicksAndRejections (internal/process/task_queues.js:93:5)

  19) node-fetch
       should allow redirected response body to be aborted:

      AssertionError: expected { type: 'system',
  name: 'FetchError',
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED',
  erroredSysCall: [Error: connect ECONNREFUSED 127.0.0.1:80] } to have property 'name' of 'AbortError', but got 'FetchError'
      + expected - actual

      -FetchError
      +AbortError
      
      at /home/akepinski/dev/node-fetch/node_modules/chai-as-promised/lib/chai-as-promised.js:302:22
      at processTicksAndRejections (internal/process/task_queues.js:93:5)

  20) node-fetch
       should remove internal AbortSignal event listener after request and response complete without aborting:
     AssertionError: expected promise to be fulfilled but it was rejected with 'FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80'
  

  21) node-fetch
       should support fetch with Request instance:
     FetchError: request to http://localhost:30001/hello failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  22) node-fetch
       should support fetch with Node.js URL object:
     FetchError: request to http://localhost:30001/hello failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  23) node-fetch
       should support fetch with WHATWG URL object:
     FetchError: request to http://localhost:30001/hello failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  24) node-fetch
       should support overwrite Request instance:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  25) node-fetch
       supports supplying a lookup function to the agent:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  26) node-fetch
       supports supplying a famliy option to the agent:
     FetchError: request to http://localhost:30001/inspect failed, reason: connect ECONNREFUSED 127.0.0.1:80
      at ClientRequest.<anonymous> (src/index.js:113:11)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

  27) Request
       should override signal on derived Request instances:
     TypeError [ERR_INVALID_URL]: Invalid URL: test
      at onParseError (internal/url.js:243:9)
      at new URL (internal/url.js:319:5)
      at new Request (src/request.js:59:16)
      at Context.<anonymous> (test/test.js:2617:26)
      at processImmediate (internal/timers.js:439:21)

  28) Request
       should allow removing signal on derived Request instances:
     TypeError [ERR_INVALID_URL]: Invalid URL: test
      at onParseError (internal/url.js:243:9)
      at new URL (internal/url.js:319:5)
      at new Request (src/request.js:59:16)
      at Context.<anonymous> (test/test.js:2629:26)
      at processImmediate (internal/timers.js:439:21)

  29) external encoding
       data uri
         should accept data uri of plain text:

      AssertionError: expected '' to equal 'Hello World!'
      + expected - actual

      +Hello World!
      
      at equal (test/test.js:2835:44)



npm ERR! Test failed.  See above for more details.
```

### CODE REVIEW

**Comment by bitinn** [comment]
Sorry for the late reply, I will look into it when available.

But the tests appear to fail to completely unrelated issues. Can we please get the CI back to a working state before proceeding to troubleshoot this issue?

https://travis-ci.org/bitinn/node-fetch/jobs/619760485

**Comment by Richienb** [comment]
@bitinn Tests are fixed in #703

**Comment by bitinn** [comment]
@xxczaki @Richienb I have re-ran the tests, the build is now passing (I don't know how `should accept data uri of plain text` test fail for the push but the PR passes), please remove utf8 package dependency and we can merge this 👍

**Comment by Richienb** [comment]
@bitinn We haven't migrated `parseUrl` to `new URL` for all of the instances yet...

**Comment by bitinn** [comment]
@xxczaki @Richienb 

I have figured out exactly why this is happening:

When we parse the URL object in `request.js`, we use `...` spread syntax to read its properties, but this doesn't work with `new URL()` object (its a completely different URL object from the ones created by `url.parse()`).

Then in `index.js` we check for `options.protocol` in order to determine the request protocol, which defaults to `https` when `protocol` property is missing. Of course a request to `http://localhost:30001/hello` using `https.request` will result in `ECONNREFUSED`

I have branch with a fix, but it's not complete, some of the tests require `new Request()` and `new Response()` to support any URL input, which `new URL` can't do without a `base` parameter. So I guess we either need to prevent these edge cases. Or use a default `base` when URL is not absolute.

I am too busy to fix these so please follow up on this problem :)

https://github.com/bitinn/node-fetch/tree/new-url-parser

**Comment by Richienb** [comment]
@bitinn It's currently late at night for me at the moment (UTC+12) so I'll take this tomorrow.

**Comment by bitinn** [comment]
Take your time, we are not in a rush here :)

Also `new URL()` relative url parsing issue is known, documented by Timothy as always: https://github.com/nodejs/node/issues/12682

**Comment by xxczaki** [comment]
Ok, so I was working with David's [branch](https://github.com/node-fetch/node-fetch/tree/new-url-parser) and here is what I did:

- To handle relative URLs I created a `parseURL` function ([inspiration](https://github.com/elastic/apm-agent-nodejs/blob/e904b2198464fe1dd0186803eb700e3403daacb4/lib/parsers.js#L220)), which is basically a wrapper around `new URL`. It uses a solution suggested [here](https://github.com/nodejs/node/issues/12682#issuecomment-377609414), which [isn't perfect](https://github.com/nodejs/node/issues/12682#issuecomment-377613017).

Now only 2 tests fail:

```
  202 passing (2s)
  2 failing

  1) node-fetch
       should reject with error if url is protocol relative:

      AssertionError: expected promise to be rejected with an error including 'Only absolute URLs are supported' but got 'Only HTTP(S) protocols are supported'
      + expected - actual

      -Only HTTP(S) protocols are supported
      +Only absolute URLs are supported
      
  

  2) Request
       should support arbitrary url:

      AssertionError: expected 'relative:///anything' to equal 'anything'
      + expected - actual

      -relative:///anything
      +anything
      
      at Context.equal (test/test.js:2735:22)
      at processImmediate (internal/timers.js:439:21)
```

---

I think we might need to release the `3.x` version with the deprecated `url.parse()`, as it looks like `new URL()` still has some important issues (like the one with relative URLs).

**Comment by Richienb** [comment]
@xxczaki In that case, we should use a ponyfill for url.parse t

**Comment by bitinn** [comment]
I have taken a brief look into the errors.

I think the first test error is easy to solve, we just need to change the error message,

Screen Shot 2020-01-23 at 12 09 58

Screen Shot 2020-01-23 at 12 10 34

https://github.com/node-fetch/node-fetch/blob/3.x/test/test.js#L113-L116

The second error I think is an edge case: `node-fetch` isn't designed with relative url in mind, due to lack of `context`. So if you pass `Request` ctor a relative url, perhaps the right thing is to throw a custom error to let users know (because it's a programming error, and a limit of `node-fetch`).

It should be possible to detect whether the url is absolute or not, and once it passes, we will use `new URL()` with a default `base` such as `relative:///`.

**Comment by xxczaki** [comment]
@bitinn I've introduced the detection of whether the URL is absolute or not (regex was taken from [here](https://github.com/sindresorhus/is-absolute-url/blob/8a57ccec064d925640860674cd32e7182eeb3596/index.js#L15)). I've also changed the error message, as you suggested.

Now, should we stop supporting relative URLs and throw a custom error instead? Or maybe we should do something else?

cc @bitinn @Richienb

**Comment by bitinn** [comment]
I think we should throw an Error stating that we don’t support relative URL yet. To me this is not possible without the “context”, so even if we workaround it, node-fetch still won’t do the same thing as browsers’ fetch.

I think we should also throw if you create a Request or Response manually with relative URL.

(From my phone)

**Comment by xxczaki** [comment]
@bitinn Ok, these tests fail:
```
1) should override signal on derived Request instances
2) should allow removing signal on derived Request instances
3) should default to null as body
4) should support arbitrary url
5) should support ArrayBuffer as body
6) should support Uint8Array as body
7) should support DataView as body
```

1, 2 and 4 are related to arbitrary URLs. 3, 5, 6 and 7 do not specify the URL (so the `Response` class assumes it's an empty string).

**Comment by bitinn** [comment]
@jimmywarting what's your thought on this? obvious we would like to support arbitrary URL but the `new URL()` just couldn't handle it without some nasty hacks.

**Comment by jimmywarting** [comment]
I also made an attempt to get in the new URL long time ago. but i also stumble upon some failing tests then.

It would be an nice addition to get in the new URL for sure. Even if it meant that we got some new failing test and possible some breaking changes. like throwing an error for constructing request with a relative  url. But that's maybe something that we can fix arbitrary URL after this.

**Comment by bitinn** [comment]
@xxczaki 

Let's go with no arbitrary url for now?

> 3, 5, 6 and 7 do not specify the URL (so the Response class assumes it's an empty string).

We can fix these by supplying a placeholder url

> 1, 2 and 4 are related to arbitrary URLs

If we can use full URL and fix these, do so; otherwise make changes to the test itself to reflect the changes

And finally, please update UPGRADE doc to reflect this, thx a lot on your efforts to move this forward!

**Comment by xxczaki** [comment]
@Richienb Yes, just removed it :smile:

**Comment by xxczaki** [comment]
Using this opportunity, I did some small refactoring:
- Split tests into different files (though the main one is still huge)
- Removed ignored xo rules to improve code quality

File: docs/v3-UPGRADE-GUIDE.md [hunk 1]
```diff
@@ -67,6 +67,10 @@ We are working towards changing body to become either null or a stream.
 
 The default user agent has been changed from `node-fetch/1.0 (+https://github.com/node-fetch/node-fetch)` to `node-fetch (+https://github.com/node-fetch/node-fetch)`.
 
+## Arbitrary URLs are no longer supported
+
+Since in 3.x we are using the WHATWG's `new URL()`, arbitrary URL parsing will fail due to lack of base.
```
Comment (bitinn) [inline docs/v3-UPGRADE-GUIDE.md]: I felt this explanation can be better, something like:

> Creating Request/Response objects with relative URLs are no longer supported
>
> We introduce Node.js core's `new URL()` API in 3.x, it offers better UTF-8 support and is WHATWG URL compatible. The drawback is, given current limit of the API (nodejs/node#12682), it's not possible to support relative URL parsing without hacks.
> Due to the lack of a browsing context on Node.js, we opted to drop support for relative URLs on Request/Response object, and it will now throw errors if you do so.
> The main `fetch()` function will support absolute URLs and data url.

Comment (xxczaki) [reply docs/v3-UPGRADE-GUIDE.md]: Agree. Just updated the upgrade guide :smile:

**Comment by xxczaki** [comment]
> * can you point me to changed tests now that there are refactoring going on, I can't see which tests are changed?

Here are the changes: https://github.com/node-fetch/node-fetch/pull/701/commits/dec40659ca25ef1ec6a7967d16c0c8b5626df38c

TL;DR:

- Deleted one test (`should support arbitrary url`)
- Updated others by supplying a placeholder URL.

Note: In the above commit I used a wrong placeholder URLs `${base}/hello`, instead of `${base}hello`. Fixed it in https://github.com/node-fetch/node-fetch/pull/701/commits/ac60e7a012f793985d674ac4fec666fd90e008e5.

**Comment by jimmywarting** [comment]
Feels like this PR is getting little out of context and being rather large and doing unrelated changes towards just changing parse with url... more to review.

**Comment by xxczaki** [comment]
@jimmywarting I can remove the refactoring commits and create a separate PR for that. However, since this is probably the last PR before the 3.x release, I wanted to include them in it.

**Comment by bitinn** [comment]
I went through the actual changes on tests and main codebase again, they look fine to me.

Travis CI show the number of tests are correct (206, vs the 207 in v3.x branch currently).

I have one final question though:

**Did we test UTF-8 URL support? The actual goal of using `new URL()`?**

**Comment by jimmywarting** [comment]
@jimmywarting I can remove the refactoring commits and create a separate PR for that. However, since this is probably the last PR before the 3.x release, I wanted to include them in it.

you don't have to but just remember that till next time

**Comment by xxczaki** [comment]
@bitinn Will something like this do the trick?

```js
it('URLs are encoded as UTF-8', () => {
	const url = 'http://example.com/möbius';

	fetch(url).then(res => expect(res.url).to.equal('http://example.com/m%C3%B6bius'));
});
```

**Comment by bitinn** [comment]
> @bitinn Will something like this do the trick?
> 
> ```js
> it('URLs are encoded as UTF-8', () => {
> 	const url = 'http://example.com/möbius';
> 
> 	fetch(url).then(res => expect(res.url).to.equal('http://example.com/m%C3%B6bius'));
> });
> ```

I think so, but better to use a local url with our test server script, if it's possible. And of course verify the test fails with current v3 branch, otherwise we are not fixing issues like #245 and #233

**Comment by xxczaki** [comment]
@bitinn Everything seems to be working fine :smile:

**Comment by bitinn** [comment]
Then I think it’s time to merge and put v3 out!

(whether we should beta it first, can be discussed in the main thread)

(From my phone)

File: src/request.js [hunk 2]
```diff
@@ -94,7 +115,7 @@ export default class Request {
 			signal = init.signal;
 		}
 
-		if (signal != null && !isAbortSignal(signal)) {
+		if (signal !== null && !isAbortSignal(signal)) {
```
Comment (SasanFarrokh) [inline src/request.js:118]: This had made some issues, if signal is undefined, throws the error

File: src/body.js [hunk 3]
```diff
@@ -293,7 +293,7 @@ export function clone(instance, highWaterMark) {
  */
 export function extractContentType(body) {
 	// Body is null or undefined
-	if (body == null) {
+	if (body === null) {
```
Comment (LinusU) [inline src/body.js:296]: The comment is no longer accurate since this now only checks for `null` (`=== null`) and doesn't include `undefined` (`== null`)

This is present in a number of places...
